### ISSUE TITLE
Task runner contract (pluggable task runners)

### ISSUE THREAD

**Issue by moltar** [projen/projen#2051]
Following up on #1826 to make a more generic issue to foster further discussion.

Specifically this comment: https://github.com/projen/projen/issues/1826#issuecomment-1211720662

> We need to make sure the contract for task runners is right. Then we can implement a default choice for the batteries-included projects based on that.

---

There are several issues at the moment with projen's task implementation:

- Extra params cannot be included (E.g. `projen test -- my.spec.ts`). This is because the projen task can include many forked tasks and thereby it is not possible to infer which sub-process needs these extra parameters.
- It sets up it's own opinion and hirerachy of tasks calling (e.g. see [build task](https://github.com/projen/projen/blob/main/docs/build.md#build)). Which makes it difficult to set up other task runners and optimize the behaviour. For example, turborepo can also run task hierarchies, but in an optimized way (using cache).

There are many competing task runners emerging. Notable ones:

- [Turborepo](https://turborepo.org/docs)
- [Wireit](https://github.com/google/wireit): #1826
- [Lage](https://github.com/microsoft/lage)

---

As a user, I want to be able to swap a task runner for a custom one. Ideally, it would be another component or package that can easily plug into the projen project instance.

Should be as simple as:

```ts
new typescript.TypeScriptProject({
  // ... projen options

  taskRunner: new CustomTaskRunner({
    // optional task runner options
  }),
})
```

Where `CustomTaskRunner` implements an interface:

```ts
class CustomTaskRunner implements IProjenTaskRunner {
  // ...
}
```

---

### Question: What should this interface look like?

```ts
interface IProjenTaskRunner {
  // ❓❓❓
}
```

**Comment by moltar** [comment]
> How are task dependencies modeled?

This is one of the main reasons why I believe the current implementation is lacking. There is no way to setup a chain of dependencies (aka "pipeline") as it is in Turborepo or Lage.

It looks like Lage and Turborepo kind-a settled on a "pipeline" interface.

- https://turborepo.org/docs/core-concepts/pipelines
- https://microsoft.github.io/lage/docs/Guide/pipeline

Maybe we can model the same way?

> How should we passing on extra parameters?

Don't have any good ideas here yet! Certainly a challenging problem.

**Comment by moltar** [comment]
> How should we passing on extra parameters?

I think we just have to accept that it is not possible to pass args to a "pipeline" task (i.e. the root task that calls other tasks).

But at the same time, projen should not block the ability to pass args to a leaf task.

The simplest case is passing args to Jest. It is very common and even natural, to do:

```sh
npm t -- tests/foot.test.ts
```

**Comment by mrgrain** [comment]
> I think we just have to accept that it is not possible to pass args to a "pipeline" task (i.e. the root task that calls other tasks).
> But at the same time, projen should not block the ability to pass args to a leaf task.

Yes but for an imagined task `test` consisting of the steps:
```
tsc
jest
upload-coverage.sh
```

I think we would still want a way to define how arguments are passed.

Ideas I can come w/o thinking to much about it are:
- adding an additional option to `exec` etc: `task.exec(command: string, shouldGetArgs = false)`
- using a producer pattern: `task.exec(command: string | (args: string[]) => string)`

**Comment by moltar** [comment]
Fair enough!

I think having the option to intercept args is a good one.

That way you can pass thru args, if you want, or you could amend them to your liking, for example, and then pass to the underlying tasks.

E.g. maybe, passing `-u` (to jest), but then some other testing framework has a different flag for the same concept (updating snapshots), which you could then overload.

**Comment by mrgrain** [comment]
For posterity: [wireit proposal](https://github.com/projen/projen/issues/1826)

**Comment by mrgrain** [comment]
I've been looking at this the last couple days. Here are my (unsorted) thoughts:

- Obviously #356 , but I have a fix for this open #2246 
- It's `TaskRuntime` that we need to swap out. The TaskModel will likely stay in projen, otherwise cross runtime compatibility will be a mess.
- Runtimes might extend the TaskModel with extra features. Using any of them will make a project less portable. That's okay as a user, probably not a good idea as a provider of a project
- We should have a separate package `projen-run` or `pjx` that contains the basic default task runtime. Advantage: Projen can be ejected from a project and tasks still work.
- Replace `spawn` with a dependency model (`dependsOn`). A simple runtime can execute tasks in order, a more complex one can parallelize and cache.
- We will likely need a definition for  input and output files, otherwise any caching attempts (wireit, nx) will be futile. However low priority.
- Tasks should be able to define more complex (positional) arguments and options
- DependsOn should be able to define calling a dependency with specific options, including passing options through from the parent. `test-task --file {testFile} --dependsOn--> jest-task {file}`
- The argument parsing might have to stay with projen in the first iteration if we want the `projen test` shortcuts. I'm not also not sure if it should be delegate to the task runtime at all, or maybe that could be something swap-able as well (but what's the point).
- Builtin tasks need to be more flexible. Especially the should be parameterized without using env vars. Currently they use a silly workaround to define env vars in the task that are then read from the builtin.
- Also 3rd party "builtin" tasks should be a thing

**Comment by mrgrain** [comment]
Related issue: #2402

**Comment by mrgrain** [comment]
I've had a play with this over the weekend. https://github.com/projen/projen/pull/2468

Lessons learned from this:
- It's straight forward to achieve this. We mostly need some sort of inversion of control/service locator thing to get the correct runtime. Otherwise it's just moving some code around.
- A Jsii Dependency Injection library would be cool.
- I think the interface would be quite simple. The version above only has a `runTask(name: string)` method as requirement. We also might want a property to list all tasks and/or find tasks by name. But this is not essential.
- Some other things will have to be explored and need to change, e.g. rendering task commands.
- We need to understand how this effects the `projen` command. When using a fully custom runtime (see below), the default `projen` command to synthesize the project won't work anymore.
- I have distilled the following patterns of Runtimes:
  - A **Fully Custom Runtime** regular projen commands won't work anymore. `.projen/tasks.json` does not exists. projen task runner would effectively be used to init the project, than delegates to the task runner.
  - A **Proxy Runtime** would change all tasks in `.projen/tasks.json` to redirect them to the actual task runner. This way the projen binary can still be used, but doesn't do the work. E.g. `projen build` would just call `nx build`. I guess users would be encouraged to use the nx commands, but both work.
  - A **Replica Runtime** keeps the original projen tasks and adds the new task runner in parallel. I don't think this is a good idea, as it might cause confusion with users and implementations might be slightly different between runtimes. A more likely variation to this would be a **HybridRuntime** that keeps the default synth task working in projen, but removes any other task,

**Comment by moltar** [comment]
> I think the interface would be quite simple.

Do you think there is value in having inputs and outputs defined in the interface (even if as optional props)?

This seems to be the way most of the build optimization tools go.

But I can see this being outside the scope of the task runner too, as long it is possible to expand on that, and be able to define these things alongside the tasks.

I am particularly thinking in the case of turborepo. It will read tasks from the `package.json` file, but it will also look at `turbo.json` to find the dependency graph of the tasks and their inputs/outputs to do the cache hashing calculations.

> A Proxy Runtime would change all tasks in .projen/tasks.json to redirect them to the actual task runner. This way the projen binary can still be used, but doesn't do the work. E.g. projen build would just call nx build. I guess users would be encouraged to use the nx commands, but both work.

Would it be possible to make this optional?

Ultimately, I don't want to have too many ways to run a command in a Node (and sub classes) project.

I would prefer to use the traditional `npm run ...` approach, and leave `projen ...` to deal with projen-specific commands.

**Comment by mrgrain** [comment]
> Do you think there is value in having inputs and outputs defined in the interface (even if as optional props)?

We should have that! But I reckon it would go into the TaskModel and a Runtime can make use of it (or not).

> But I can see this being outside the scope of the task runner too, as long it is possible to expand on that, and be able to define these things alongside the tasks.

I briefly looked into swapping the TaskModel as well. But that becomes much more complicated since it quickly ends up being a nightmare in terms of support. You can't develop a universal component the depends on a specific model. Or rather you can, but that's just the same situation we are currently in. So it seems better to extend the current model with common patterns.

> I am particularly thinking in the case of turborepo. It will read tasks from the `package.json` file, but it will also look at `turbo.json` to find the dependency graph of the tasks and their inputs/outputs to do the cache hashing calculations.

> > A Proxy Runtime would change all tasks in .projen/tasks.json to redirect them to the actual task runner. This way the projen binary can still be used, but doesn't do the work. E.g. projen build would just call nx build. I guess users would be encouraged to use the nx commands, but both work.
> 
> Would it be possible to make this optional?

That's what I'm trying to figure out. If we were to go with the interface I have on the branch, it would be 100% up to the replacing runtime to decide what to support. To be clear, I'm not sure a proxy runtime is a good idea. This is just something that I discovered while playing around with it. One reason for it could be to have compatibility with something the projen vscode extension.

> Ultimately, I don't want to have too many ways to run a command in a Node (and sub classes) project.
> 
> I would prefer to use the traditional `npm run ...` approach, and leave `projen ...` to deal with projen-specific commands.

And that's a fair requirement. Mind you though, there are no projen-specific commands. The default tasks is just a fancy way to execute `node .projenrc.js`. It's totally possible to have all commands define in `package.json`.

For compatibility reasons with other languages it makes a whole of sense that projen ships with a task runtime though.

**Comment by moltar** [comment]
Now that we have constructs-based projen, would modelling tasks thru constructs be a cleaner solution? Could this be a whole new way to declare tasks that we could opt into, but also be compatible with the old Task class?

I think having a clean constructs-based interface does represent the build graph quite well. Dependencies become implicit via scoping, but can also be setup via the addDependency mechanism, if one needs to have that be explicit.

**Comment by mrgrain** [comment]
That would definitely be the better way! Although we still need an implementation of `Dependencies`, it's currently not used.

**Comment by moltar** [comment]
> Although we still need an implementation of `Dependencies`, it's currently not used.

Is it the #3056 issue?

**Comment by mrgrain** [comment]
> > Although we still need an implementation of `Dependencies`, it's currently not used.
> 
> Is it the #3056 issue?

Yes that covers dependencies, amongst other things.

**Comment by cameroncf** [comment]
Adding some notes here after having a conversation with @moltar last week, primarily regarding my own mental model for task orchestration. I currently use the native Projen task model in a monorepo environment and then layer on TurboRepo strictly for cross project orchestration and (remote) output caching. This allows me to do normal task stuff in Projen without much pain and then I can chain and Parallelize them using Turbo. The turbo.json dependency graph is built and configured by each sub-project as they are added. This centralizes complex dependency configuration in one spot while still delegating inputs/outputs to individual project types.

- I have two places that I call turbo, the root project's pre-compile and post-compile tasks.
- I run `pnpm build` on the root project as normal but there is virtually nothing to build in the root project.
- Once the root build hits post-compile, turbo takes over.
- I don't run "build" in any sub-projects, only individual tasks orchestrated by Turborepo. 
- I have a lot of "subproject#nuxt:build" and other custom tasks in subprojects that may be non standard. 

So conceptually, if we remove Turbo from the concept, I'm splitting apart the task definitions from the orchestration configuration.  If projen went this route, the tasks.json could stay the same as it is today, but all the spawns and cross project task orchestration could simply be abstracted elsewhere. I'm not sure tis is a better idea than other ideas which have been proposed already but I wanted to add these thoughts in case they prompt more ideas.

**Comment by mrgrain** [comment]
Thanks @cameroncf ! Appreciate this a lot. I strictly view `task.json` as an artifact for the built-in projen Taks Runner. It's what `scripts` in `package.json` is for `npm`, `yarn` & `pnpm`. Or `nx.json` for Nx.

I agree that orchestration ob mono-repo tasks dependency and instructions how to run tasks are two different things. 

In my head all of the following should be possible:
- A projen task runner that uses `task.json` and does not put anything in `package.json`
- A Yarn-only task runner that cannot be called from projen, forcing users to run `yarn my-task` only.
- A Nx task runner that cannot be called from projen and does not put anything in `package.json` or `task.json`
- A projen task runner that has its commands mirrored to `package.json`
-  A Nx task runner that has it's commands mirrored to `package.json`

Most importantly, for any projects using task runners other than projen, projen would never be called. It would simply not exists as far as the user is concerned. Instead you would call `yarn default` or `nx default`. Although we might want to keep a stub functionality around in the projen binary to at least run the default task and/or provide an error message. It could also work as a proxy task runner, like `package.json`

**Comment by cameroncf** [comment]
I'm putting this comment here instead of in PR #3162 so that it can be added to this discussion.

I've spent quite a built of time making Projen work with TurboRepo (in a private repository, for now) and after building and re-building my ideas for some time, I have come to the following mental model:

- **Tasks should be Atomic**  In order for task inputs and outputs to be cachable by an outside tool (such as TurboRepo) they need to be atomic. I don't even use Projen's `build` task anymore because I want to cache things like `compile`, `test`, and `package` separately using different input and output patterns. The fact that the Projen `test` task (by default) does both Unit Tests and eslint is also slightly bad since it's not quite as atomic as I would like for capability.
- **Separate Tasks and Orchestration** - Once all my Projen tasks are atomic, I turn over orchestrating them to TurboRepo. This means that whatever Projen normally puts into the `build` task, I put into TurboRepo's configuration. 
- **Monorepo Root is an Exception** - In my case, I am working with a Monorepo containing many packages and CDK projects. I still allow the normal Projen `build` task to run in the normal Projen way in the root of the monorepo.  This is because it's outputs are spread out across other projects, and also because it may be changing the Turborepo config file when it runs and I want to avoid circular dependancies or other problems that could create. After that root projen build is complete, I hand everything else over to Turbo. This PJ-then-Turbo task in my system is called `build:all` so that I don't step on any existing tasks and can switch it out with another orchestration engine should I choose to do so.

In summary this means (for my use case and based on my experiences), any sort of "run before / after" task dependancies or orchestrations should be reserved for the outside task runner like Turborepo. I don't have a lot of experience with Nx or other tooling so there could be other reasons or use cases that I am blind to.

I'd really like to extract my work into a public facing repo at some point just as a proof of concept for further comment by the community and to help build this RFC. Maybe, maybe, maybe....

**Comment by mrgrain** [comment]
> Tasks should be Atomic In order for task inputs and outputs to be cachable by an outside tool (such as TurboRepo) they need to be atomic. I don't even use Projen's build task anymore because I want to cache things like compile, test, and package separately using different input and output patterns. The fact that the Projen test task (by default) does both Unit Tests and eslint is also slightly bad since it's not quite as atomic as I would like for capability.

I mostly agree with this. And I don't think it was the intention of https://github.com/projen/projen/pull/3162 to be not atomic. But we also have a migration problem to deal with. projen is doing it a certain way today, and we need to start moving away from this.

I do want to call out that outside task runners still have a concept of requirements. The `test` target, requires the `compile` target to be completed first. But it can be up to the outside task runner to decide if `compile` needs to run again at all, if the current output can be used (because no input has changed), or if a cached output should be used.

So I agree that the concept of `pre-compile` in projen core is bad.

**Comment by cameroncf** [comment]
> But we also have a migration problem to deal with. projen is doing it a certain way today, and we need to start moving away from this.

The way I have approached this is to keep the projen task runner exactly how it is and simply add Turbo as an orchestration layer on top (Turbo still executing the underlying projen tasks as they exist today). Backward compatible with very little change needed in Projen core. This approach seems the safest to me since each runner (NX, Turbo, etc) will have it's own way of doing things and it's going to be very difficult to come up with a lowest common denominator Interface for this within Projen itself.

I love to be wrong though and I don't know a ton about other task runners. I am very interested in hearing how others see the problem and potential solutions.
